File: src/scrapers/error_handler.py
from functools import wraps
from typing import TypeVar, ParamSpec
from collections.abc import Callable

import time
import logging
import requests


P = ParamSpec("P")
R = TypeVar("R")

logger = logging.getLogger(__name__)


def with_retries(max_retries: int = 3, delay: float = 0.1, func_name: str | None = None):
    def decorator(func: Callable[P, R]):
        @wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R | None:
            for_args_str = f"for {', '.join(str(arg) for arg in args)}"
            error_log = (
                f"Error getting {func_name} {for_args_str}" if func_name else f"Error for {args}"
            )

            for attempt in range(max_retries + 1):
                try:
                    return func(*args, **kwargs)

                except requests.exceptions.HTTPError as e:
                    code = e.response.status_code
                    reason = e.response.reason

                    if code == 429 and attempt < max_retries:
                        logger.warning(
                            f"Ratelimited {for_args_str}. Retrying in {delay} seconds..."
                        )
                        time.sleep(delay)
                        continue

                    else:
                        logger.error(f"{error_log}: {reason} [{code}]")
                        return None

                except requests.exceptions.RequestException as e:
                    if attempt < max_retries:
                        logger.warning(f"{error_log}: {e}. Retrying in {delay} seconds...")
                        time.sleep(delay)
                        continue
                    else:
                        logger.error(f"{error_log} after {max_retries} retries: {e}")
                        return None

                except Exception as e:
                    logger.error(f"{error_log}: {e}")
                    return None

            logger.error(f"{error_log} after {max_retries} retries")
            return None

        return wrapper

    return decorator

File: src/scrapers/error_handler.ts
export function withRetries(max_retries: number = 3, delay: number = 100) {
  function decorator<TArgs extends unknown[], TReturn extends Promise<unknown>>(
    func: (...args: TArgs) => TReturn,
  ) {
    async function wrapper(...args: TArgs): Promise<TReturn | null> {
      const for_args_str = `for ${args.join(", ")}` as const;
      const error_log = func.name
        ? (`Error in ${func.name} ${for_args_str}` as const)
        : (`Error ${for_args_str}` as const);
      for (let attempt = 0; attempt <= max_retries; attempt++) {
        try {
          return func(...args);
        } catch (e) {
          if (attempt < max_retries) {
            console.warn(`${error_log}: ${e}. Retrying in ${delay} seconds...`);
            await new Promise((resolve) => setTimeout(resolve, delay));
            continue;
          } else {
            console.error(`${error_log} after ${max_retries} retries: ${e}`);
            return null;
          }
        }
      }
      console.error(`${error_log} after ${max_retries} retries`);
      return null;
    }
    return wrapper;
  }
  return decorator;
}
